modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,327 @@
+  readonly value: T;
+}
+
+// Not exported, instead duplicated in Maybe.ts as an identical type to make
+// imports nice for consumers. Redefined here to avoid looped imports.
+type Maybe<T> = Nothing | Just<T>;
+
+// --- Identities ---
+
+// Just a small optimisation - all nothing objects are identical, why not make
+// them point to the same object?
+const singletonNothing: Nothing = { tag: Tag.Nothing };
+
+// Create a new Maybe<T> that **does not** contain a value. T can be any type
+// (and defaults to the literal any type) because we don't actually have a T.
+export const nothing = <T = any>(): Maybe<T> => singletonNothing;
+
+// Creates a new Maybe<T> that **does** contain a value.
+export const just = <T>(value: T): Maybe<T> => ({
+  tag: Tag.Just,
+  value
+});
+
+// Perform a transform on a Maybe type. Similar to the fold method on a number
+// of ADTs in the TinyMCE codebase, you pass two functions into this - if `self`
+// **does not** contain a value then `onNothing` will be called (with no
+// argument) but if `self` **does** contain a value then `onJust` will be
+// called (with the value as its argument).
+//
+// For the FP enthusiasts in the room, this function:
+// 1. Could be used to implement all of the functions below (although it isn't)
+// 2. Forms a catamorphism
+export const fold = <T, U>(onNothing: () => U, onJust: (value: T) => U) => (self: Maybe<T>): U => {
+  switch (self.tag) {
+    case Tag.Nothing:
+      return onNothing();
+    case Tag.Just:
+      return onJust(self.value);
+  }
+};
+
+// Determine if this Maybe object **does not** contain a value.
+export const isNothing = <T>(self: Maybe<T>): self is Nothing => self.tag === Tag.Nothing;
+
+// Determine if this Maybe object contains a value. If this function returns
+// true, you can access the value using `.value` on the object directly.
+export const isJust = <T>(self: Maybe<T>): self is Just<T> => self.tag === Tag.Just;
+
+// --- Functor ---
+
+// Perform a transform on a Maybe object, **if** there is a value. If you
+// provide a function to turn a T into a U, this is the function you use to turn
+// a `Maybe<T>` into a `Maybe<U>`. If self **does** contain a value then the
+// output will also contain a value (that value being the output of
+// `mapper(self.value)`), and if self **does not** contain a value then
+// neither will the output.
+export const map = <T, U>(mapper: (value: T) => U) => (self: Maybe<T>): Maybe<U> =>
+  isJust(self) ? just(mapper(self.value)) : nothing<U>();
+
+// --- Applicative ---
+
+// Perform a transform on two Maybe objects, **if** they both have values. If
+// either of the objects does not contain a value, this will return an empty
+// Maybe (nothing), but if all of the input objects do have a value then `fn`
+// will be called (with all of those values as its arguments) and the output
+// will be wrapped in another Maybe.
+export const lift2 = <A, B, Out>(a: Maybe<A>, b: Maybe<B>, fn: (a: A, b: B) => Out): Maybe<Out> => {
```

Comment:
Worth noting our normal convention is to use `R` here generally instead of `Out`.